src/rosdistro/manifest_provider/cache.py

```diff
@@ -32,6 +32,35 @@
 # POSSIBILITY OF SUCH DAMAGE.
 
 from rosdistro import logger
+from xml.dom import minidom
+
+
+def sanitize_xml(xml_string):
+    """ Returns a version of the supplied XML string with comments and all whitespace stripped,
+        including runs of spaces internal to text nodes. The returned string will be encoded,
+        so str (Python 2) or bytes (Python 3). """
+    def _squash(node):
+        drop_nodes = []
+        for x in node.childNodes:
+            if x.nodeType == minidom.Node.TEXT_NODE:
+                if x.nodeValue:
+                    x.nodeValue = ' '.join(x.nodeValue.strip().split())
+            elif x.nodeType == minidom.Node.ELEMENT_NODE:
+                _squash(x)
+            elif x.nodeType is minidom.Node.COMMENT_NODE:
+                drop_nodes.append(x)
+        for x in drop_nodes:
+            node.removeChild(x)
+        return node
+    try:
+        # Python 2. The minidom module parses as ascii, so we have to pre-encode.
+        if isinstance(xml_string, unicode):
+            xml_string = xml_string.encode('utf-8')
+        # Returns an encoded str.
+        return _squash(minidom.parseString(xml_string)).toxml('utf-8')
+    except NameError:
+        # Python 3 returns a unicode str.
+        return _squash(minidom.parseString(xml_string)).toxml()
```

Comment:
Done, although AttributeError on encode is not adequate to detect Py2, since the `str` type in Python 2 _has_ an `encode` method which fails in other ways. I believe what I have now works and is in line with your suggestion here.